crates/networking/p2p/src/network.rs

```diff
@@ -225,6 +279,52 @@ impl Network {
             }
         }
     }
+
+    fn upsert_peer_status(&self, peer_id: PeerId, status: PeerStatus) {
+        if let Ok(mut peers) = self.peers.write() {
+            // todo: PeerId and string mismatch
+            peers.insert(
+                peer_id,
+                Peer {
+                    id: peer_id.to_string(),
+                    status,
+                },
+            );
+            info!("Peer {}: status changed to {:?}", peer_id, status);
+        }
+    }
+
+    // todo: when should a peer be removed?
+    // fn remove_peer(&self, peer_id: &PeerId) {
+    //     if let Ok(mut peers) = self.peers.write() {
+    //         peers.remove(peer_id);
+    //         info!("Peer {}: removed from tracking", peer_id);
+    //     }
+    // }
+
+    pub fn get_peer_counts(&self) -> (usize, usize, usize, usize) {
+        if let Ok(peers) = self.peers.read() {
+            let connected = peers
+                .values()
+                .filter(|&peer| peer.status == PeerStatus::Connected)
+                .count();
+            let connecting = peers
+                .values()
+                .filter(|&peer| peer.status == PeerStatus::Connecting)
+                .count();
+            let disconnecting = peers
+                .values()
+                .filter(|&peer| peer.status == PeerStatus::Disconnecting)
+                .count();
+            let disconnected = peers
+                .values()
+                .filter(|&peer| peer.status == PeerStatus::Disconnected)
+                .count();
+            (connected, connecting, disconnecting, disconnected)
+        } else {
+            (0, 0, 0, 0) // Default in case of lock error
+        }
+    }
```

Comment:
```suggestion
    pub fn get_peer_counts(&self) -> (usize, usize, usize, usize) {
        let mut connected = 0;
        let mut connecting = 0;
        let mut disconnecting = 0;
        let mut disconnected = 0;    
        if let Ok(peers) = self.peers.read() {
            for peer in peers {
                match peer.status {
                     PeerStatus::Connected => connected += 1,
                     PeerStatus::Connecting => connecting += 1,
                     PeerStatus::Disconnecting => disconnecting += 1,
                     PeerStatus::Disconnected => disconnected += 1,
                }
            }
        }
        (connected, connecting, disconnecting, disconnected)
    }
```

I think something like this would be better as the pervous approach looks for peers 4 times